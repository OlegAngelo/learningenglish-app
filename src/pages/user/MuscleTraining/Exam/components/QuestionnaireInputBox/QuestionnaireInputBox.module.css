.response {
  -webkit-text-stroke: .55px #FFFFFF;
}

.inputBox {
  padding-left: 11.5px;
  padding-right: 11.5px;
  border: 1px solid #FFFFFF;
  box-shadow: 0px 2px 6px rgba(0, 0, 0, 0.1);
}

.correctOutline {
  border: 2px solid #03DAC6;
}

.incorrectOutline {
  border: 2px solid #E34E42;
}

.textField {
  font: 2.5ch monospace;
}

input:focus {
  outline: none;
}

.red {
  color: #E34E42;
}

.inputFontSize {
  font-size: 4vw !important;
}

.inputWidth {
  width: 3.7vw !important;
}

.inputMargin {
  margin-left: 7px !important;
  margin-right: 7px !important;
}

/* ----------- iPhone 6, 6S, 7 and 8 ----------- */
@media only screen
  and (max-device-width: 375px)
  and (min-device-height: 667px)
  and (-webkit-min-device-pixel-ratio: 2)
  and (orientation: portrait) {

  .softKeyboardWrapper {
    position: relative;
    margin-top: 5%;
    margin-bottom: 20%;
  }

}

/* ----------- iPhone 6+, 7+ and 8+ ----------- */
@media only screen
  and (max-device-width: 414px)
  and (min-device-height: 736px)
  and (-webkit-min-device-pixel-ratio: 3)
  and (orientation: portrait) {

  .softKeyboardWrapper {
    position: relative;
    margin-top: 3%;
    margin-bottom: 20%;
  }

}

/* ----------- iPhone X ----------- */
@media only screen
  and (max-device-width: 375px)
  and (min-device-height: 812px)
  and (-webkit-min-device-pixel-ratio: 3)
  and (orientation: portrait) {

  .softKeyboardWrapper {
    bottom: 8%;
    position: fixed;
  }
}

/* ----------- iPhone 5, 5S, 5C and 5SE ----------- */
@media only screen
  and (max-device-width: 320px)
  and (min-device-height: 568px)
  and (-webkit-min-device-pixel-ratio: 2)
  and (orientation: portrait) {
    .softKeyboardWrapper {
      position: relative;
      margin-top: 14%;
      margin-bottom: 5%;
    }

    .inputFontSize {
      font-size: 3vw !important;
    }
  
    .inputWidth {
      font-size: 3vw !important;
    }
  
    .inputMargin {
      margin-left: 2px !important;
      margin-right: 2px !important;
    }
  }


/* ----------- iPhone 12, 12 Pro ----------- */
@media only screen
  and (max-device-width: 390px)
  and (max-device-height: 844px)
  and (-webkit-min-device-pixel-ratio: 3)
  and (orientation: portrait) {
    .softKeyboardWrapper {
      position: relative;
      margin-top: 8%;
    }
  }
