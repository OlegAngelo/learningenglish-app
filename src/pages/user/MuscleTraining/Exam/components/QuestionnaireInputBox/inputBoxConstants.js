import style from "./QuestionnaireInputBox.module.css";

const inputBoxConstants = {
  input: {
    type: 'input',
    response: '',
    classPrefix: '',
  },
  correct: {
    type: 'correct',
    response: 'Excellent!!',
    classPrefix: 'correct',
  },
  incorrect: {
    type: 'incorrect',
    response: 'keep it up!',
    classPrefix: "incorrect",
  },

  inputBoxStyle: function (wordLength, order = '') {
    const charWidth = 1;
    const gap = 0.5;
    let inWidth = 1 * (charWidth + gap);
    let styles = {
      fontSize: '24px',
      width: '24px',
    };

    if (wordLength >= 13) {
      styles = {
        fontSize: '16px',
        width: '14px',
      };
    } else if (wordLength >= 10) {
      styles = {
        fontSize: '20px',
        width: '18px',
      };
    }

    return {
      fontSize: styles.fontSize,
      width: `${inWidth - 0.8}ch`,
      borderBottom: '1px solid #C4C4C4',
      width: styles.width,
      textAlign: 'center',
      marginLeft: order === 'first' ? '10px' : '3px',
      marginRight: order === 'last' ? '10px' : '3px',
      fontFamily: 'SF Pro Text',
      lineHeight: '30px',
    };
  },

  tagStyle: function (wordLength) {
    let styles = {
      fontSize: '24px',
      width: '24px',
    };

    if (wordLength >= 13) {
      styles = {
        fontSize: '16px',
        width: '14px',
      };
    } else if (wordLength >= 10) {
      styles = {
        fontSize: '20px',
        width: '18px',
      };
    }

    return {
      fontSize: styles.fontSize,
      width: styles.width,
      textAlign: 'center',
      marginRight: '5px',
      fontFamily: 'SF Pro Text',
      lineHeight: '30px',
    };
  },

  prepositionStyle: function (index) {
    return {
      fontSize: '24px',
      letterSpacing: '8px',
      textAlign: 'center',
      marginLeft: '8px',
      marginRight: '8px',
      fontFamily: 'SF Pro Text',
      lineHeight: '30px',
    };
  },

  inputBoxClass: function (maxLength) {
    if (maxLength <= 13) return;
    return `${style.inputFontSize} ${style.inputWidth}`;
  },

  prepositionClass: function (maxLength) {
    if (maxLength <= 13) return;
    return `${style.inputFontSize} ${style.inputMargin}`;
  },
};

export default inputBoxConstants;
