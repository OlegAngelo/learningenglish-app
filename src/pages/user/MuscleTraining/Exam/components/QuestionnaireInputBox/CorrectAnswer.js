import React, { Fragment } from 'react';

import inputBoxConstants from './inputBoxConstants';

import styles from './QuestionnaireInputBox.module.css';

const CorrectAnswerField = ({
  answer,
  wordLength,
  hint,
  typing,
  correctAnswer,
  learningType,
  tagIdentifier,
  tagValues,
  answerWithNoTag,
}) => {
  const typingArray = typing.split('');
  const typingLength = typing.length;
  const hintIndices = () => {
    const hintArray = hint.split('');
    const indices = [];

    hintArray.forEach((char, index) => {
      if (char.toLowerCase() === answer[index]) {
        indices.push(index);
      }
    });

    return indices;
  };

  const isHint = (index) => {
    return hintIndices().includes(index);
  };
  
  const displayAnswer = (index = null) => {
    let answerArray = answer.split("");

    return answerArray.map((char, i) => {
      let order = i === 0 ? 'first' : i === answerArray.length - 1 ? 'last' : 'between';

      return (
        <span
          key={index ? `${index}${i}` : i}
          className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isHint(i) ? styles.red : ''}`}
          style={inputBoxConstants.inputBoxStyle(wordLength, order)}
        >
          {char}
        </span>
      );
    });
  };

  const displayAnswerWithTagsForWords = (titleArray, correctAnswerIndex, updatedWordLength) => {
    return titleArray.map((preposition, index) => {
      let items = [];

      if ((correctAnswerIndex == 0 && index == correctAnswerIndex) || index == correctAnswerIndex) {
        items.push(tagPositioning(updatedWordLength));
      } else {
        items.push([
          <span
            className={`inline-block ${inputBoxConstants.prepositionClass(typingLength)}`}
            key={index}
            style={inputBoxConstants.prepositionStyle(index)}
          >
            {preposition}
          </span>
        ]);
      }
      return items;
    });
  };

  const tagPositioning = (updatedWordLength) => {
    if (tagIdentifier.tagType === 'start') {
      return (
        <Fragment>
          {displayTagValues()}
          {[...Array(updatedWordLength)].map((a, i) => {
            return (
              <span
                key={i}
                className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isHint(tagIdentifier.tagTotal + i) ? styles.red : ''}`}
                style={inputBoxConstants.inputBoxStyle(wordLength)}
              >
                {answerWithNoTag[i]}
              </span>
            );
          })}
        </Fragment>
      );
    }
    if (tagIdentifier.tagType === 'end') {
      return (
        <Fragment>
          {[...Array(updatedWordLength)].map((a, i) => {
            return (
              <span
                key={i}
                className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isHint(i) ? styles.red : ''}`}
                style={inputBoxConstants.inputBoxStyle(wordLength)}
              >
                {answerWithNoTag[i]}
              </span>
            );
          })}
          {displayTagValues()}
        </Fragment>
      );
    }
    if (tagIdentifier.tagType === 'middle') {
      return (
        <Fragment>
          {displayPrevTagLines()}
          {displayTagValues()}
          {displayAfterTagLines()}
        </Fragment>
      );
    }
  };

  const displayTagValues = () => {
    return tagValues.map((t, i) => {
      return (
        <span
          className={`inline-block ${inputBoxConstants.inputBoxClass(typingLength)}`}
          style={inputBoxConstants.tagStyle(wordLength)}
          key={i}
        >
          {t.value}
        </span>
      );
    });
  };

  const displayPrevTagLines = () => {
    let lines = [];
    for (let index = 0; index < typingArray.length; index++) {
      if (index > tagIdentifier.prevIndex) break;
      lines.push(
        <span
          key={index}
          className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isHint(index) ? styles.red : ''}`}
          style={inputBoxConstants.inputBoxStyle(wordLength)}
        >
          {answerWithNoTag[index]}
        </span>
      );
    }
    return lines.map((line, i) => {
      return line;
    });
  };

  const displayAfterTagLines = () => {
    let lines = [];
    let ansCounter = tagIdentifier.prevIndex + 1;
    let wordsCount = typing.split(' ').length;
    let wordArray = (wordsCount == 1) ? typing.split('') : correctAnswer.split('');
    for (let index = 0; index < wordArray.length; index++) {
      if (index > tagIdentifier.prevIndex + tagIdentifier.tagTotal) {
        lines.push(
          <span
            key={ansCounter}
            className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isHint(index) ? styles.red : ''}`}
            style={inputBoxConstants.inputBoxStyle(wordLength)}
          >
            {answerWithNoTag[ansCounter]}
          </span>
        );
        ansCounter++;
      }
    }
    return lines.map((line, i) => {
      return line;
    });
  };

  const displayItems = () => {
    const titleArray = typing.toLowerCase().split(' ');
    const correctAnswerIndex = titleArray.indexOf(correctAnswer.toLowerCase());

    // For phrases and words that has no tags and has no prepositions
    if ((typing.toLowerCase() === correctAnswer.toLowerCase() && learningType == 'word') && !tagIdentifier) {
      return displayAnswer();
    }

    const updatedWordLength = (tagIdentifier) ? wordLength - tagIdentifier.tagTotal : 0;
    // For phrases that have tags
    if (tagIdentifier && learningType == 'phrase') {
      return tagPositioning(updatedWordLength);
    }
    // For words that have tags && prepositions OR tags only (1 word or more)
    if (tagIdentifier && learningType == 'word') {
      return displayAnswerWithTagsForWords(titleArray, correctAnswerIndex, updatedWordLength);
    }

    // For words that has no tags but has prepositions
    return titleArray.map((preposition, index) => {
      let items = [];
      let order =
        index === 0 ? 'first' : index === titleArray.length - 1 ? 'last' : 'between';

      if ((correctAnswerIndex == 0 && index == correctAnswerIndex) || index == correctAnswerIndex) {
        items.push(displayAnswer(index));
      } else {
        items.push([
          <span
            className={`inline-block ${inputBoxConstants.prepositionClass(typingLength)}`}
            key={index}
            style={inputBoxConstants.prepositionStyle(index, order)}
          >
            {preposition}
          </span>
        ]);
      }
      return items;
    });
  };

  return (
    <div className={`text-basic-100 ${styles.textField} flex justify-center`}>
      {displayItems()}
    </div>
  );
};

export default CorrectAnswerField;
