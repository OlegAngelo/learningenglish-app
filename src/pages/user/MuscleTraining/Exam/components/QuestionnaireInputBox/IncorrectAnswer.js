import React, { useEffect, useState, Fragment } from 'react';

import inputBoxConstants from './inputBoxConstants';

import styles from './QuestionnaireInputBox.module.css';

const IncorrectAnswerField = ({
  answer,
  wordLength,
  incorrectIndices,
  hint,
  typing,
  correctAnswer,
  learningType,
  tagIdentifier,
  tagValues,
  answerWithNoTag,
}) => {
  const typingArray = typing.split('');
  const typingLength = typing.length;
  const hintIndices = () => {
    const hintArray = hint.split('');
    const indices = [];

    hintArray.forEach((char, index) => {
      if (char === answer[index]) {
        indices.push(index);
      }
    });

    return indices;
  };

  const isIncorrectOrHint = (index) => {
    return incorrectIndices.includes(index) || hintIndices().includes(index);
  };
  
  const displayAnswer = (index = null, updatedWordLength = null) => {
    const finalWordLength = !updatedWordLength ? wordLength : updatedWordLength;
    return [...Array(finalWordLength)].map((e, i) => {
      let order = i === 0 ? 'first' : i === finalWordLength - 1 ? 'last' : 'between';

      return (
        <span
          key={index ? `${index}${i}` : i}
          style={inputBoxConstants.inputBoxStyle(wordLength, order)}
          className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isIncorrectOrHint(i) ? styles.red : ''}`}
        >
          {answer[i] ? answer[i] : `\xa0`}
        </span>
      );
    });
  };

  const displayAnswerWithTagsForWords = (titleArray, correctAnswerIndex, updatedWordLength) => {
    return titleArray.map((preposition, index) => {
      let items = [];
      if ((correctAnswerIndex == 0 && index == correctAnswerIndex) || index == correctAnswerIndex) {
        items.push(tagPositioning(updatedWordLength));
      } else {
        items.push([
          <span
            className={`inline-block ${inputBoxConstants.prepositionClass(typingLength)}`}
            key={index}
            style={inputBoxConstants.prepositionStyle(index)}
          >
            {preposition}
          </span>
        ]);
      }
      return items;
    });
  };

  const tagPositioning = (updatedWordLength) => {
    if (tagIdentifier.tagType === 'start') {
      return (
        <Fragment>
          {displayTagValues()}
          {[...Array(updatedWordLength)].map((a, i) => {
            return (
              <span
                key={i}
                style={inputBoxConstants.inputBoxStyle(wordLength)}
                className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isIncorrectOrHint(tagIdentifier.tagTotal + i) ? styles.red : ''}`}
              >
                {answerWithNoTag[i] ? answerWithNoTag[i] : `\xa0`}
              </span>
            );
          })}
        </Fragment>
      );
    }
    if (tagIdentifier.tagType === 'end') {
      return (
        <Fragment>
          {[...Array(updatedWordLength)].map((a, i) => {
            return (
              <span
                key={i}
                style={inputBoxConstants.inputBoxStyle(wordLength)}
                className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isIncorrectOrHint(i) ? styles.red : ''}`}
              >
                {answerWithNoTag[i] ? answerWithNoTag[i] : `\xa0`}
              </span>
            );
          })}
          {displayTagValues()}
        </Fragment>
      );
    }
    if (tagIdentifier.tagType === 'middle') {
      return (
        <Fragment>
          {displayPrevTagLines()}
          {displayTagValues()}
          {displayAfterTagLines()}
        </Fragment>
      );
    }
  };

  const displayTagValues = () => {
    return tagValues.map((t, i) => {
      return (
        <span
          className={`inline-block ${inputBoxConstants.inputBoxClass(typingLength)}`}
          style={inputBoxConstants.tagStyle(wordLength)}
          key={i}
        >
          {t.value}
        </span>
      );
    });
  };

  const displayPrevTagLines = () => {
    let lines = [];
    for (let index = 0; index < typingArray.length; index++) {
      if (index > tagIdentifier.prevIndex) break;
      lines.push(
        <span
          key={index}
          style={inputBoxConstants.inputBoxStyle(wordLength)}
          className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isIncorrectOrHint(index) ? styles.red : ''}`}
        >
          {answerWithNoTag[index] ? answerWithNoTag[index] : `\xa0`}
        </span>
      );
    }
    return lines.map((line, i) => {
      return line;
    });
  };

  const displayAfterTagLines = () => {
    let lines = [];
    let ansCounter = tagIdentifier.prevIndex + 1;
    let wordsCount = typing.split(' ').length;
    let wordArray = (wordsCount == 1) ? typing.split('') : correctAnswer.split('');
    for (let index = 0; index < wordArray.length; index++) {
      if (index > tagIdentifier.prevIndex + tagIdentifier.tagTotal) {
        lines.push(
          <span
            key={index}
            style={inputBoxConstants.inputBoxStyle(wordLength)}
            className={`${inputBoxConstants.inputBoxClass(typingLength)} ${isIncorrectOrHint(index) ? styles.red : ''}`}
          >
            {answerWithNoTag[ansCounter] ? answerWithNoTag[ansCounter] : `\xa0`}
          </span>
        );
        ansCounter++;
      }
    }
    return lines.map((line, i) => {
      return line;
    });
  };

  const displayItems = () => {
    const titleArray = typing.toLowerCase().split(' ');
    const correctAnswerIndex = titleArray.indexOf(correctAnswer.toLowerCase());

    // For phrases and words that has no tags and has no prepositions
    if ((typing.toLowerCase() === correctAnswer.toLowerCase() && learningType == 'word')  && !tagIdentifier) {
      return displayAnswer();
    }

    const updatedWordLength = (tagIdentifier) ? wordLength - tagIdentifier.tagTotal : 0;
    // For phrases that have tags
    if (tagIdentifier && learningType == 'phrase') {
      return tagPositioning(updatedWordLength);
    }
    // For words that have tags && prepositions OR tags only (1 word or more)
    if (tagIdentifier && learningType == 'word') {
      return displayAnswerWithTagsForWords(titleArray, correctAnswerIndex, updatedWordLength);
    }

    // For words that has no tags but has prepositions
    return titleArray.map((preposition, index) => {
      let items = [];
      let order =
        index === 0 ? 'first' : index === titleArray.length - 1 ? 'last' : 'between';

      if ((correctAnswerIndex == 0 && index == correctAnswerIndex) || index == correctAnswerIndex) {
        items.push(displayAnswer(index, correctAnswer.length));
      } else {
        items.push([
          <span
            className={`inline-block ${inputBoxConstants.prepositionClass(typingLength)}`}
            key={index}
            style={inputBoxConstants.prepositionStyle(index, order)}
          >
            {preposition}
          </span>
        ]);
      }
      return items;
    });
  };

  return (
    <div className={`text-basic-100 flex justify-center ${styles.textField}`}>
      {displayItems()}
    </div>
  );
};

export default IncorrectAnswerField;
