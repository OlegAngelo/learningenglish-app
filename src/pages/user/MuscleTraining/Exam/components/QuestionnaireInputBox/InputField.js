import React, { useEffect, Fragment } from 'react';
import { isMobileSafari } from 'react-device-detect';

import inputBoxConstants from './inputBoxConstants';

const InputField = ({
  wordLength,
  answer,
  setIsShowfooter,
  typing,
  correctAnswer,
  learningType,
  tagIdentifier,
  tagValues,
}) => {
  const typingArray = typing.split('');
  const typingLength = typing.length;

  useEffect(() => {
    if (isMobileSafari) {
      setIsShowfooter(true);
    }
  },[]);

  const displayLines = (index = null, updatedWordLength = null) => {
    const finalWordLength = !updatedWordLength ? wordLength : updatedWordLength;
    return [...Array(finalWordLength)].map((e, i) => {
      let order = i === 0 ? 'first' : i === finalWordLength - 1 ? 'last' : 'between';

      return (
        <span
          className={`inline-block ${inputBoxConstants.inputBoxClass(typingLength)}`}
          key={index ? `${index}${i}` : i}
          style={inputBoxConstants.inputBoxStyle(wordLength, order)}
        >
          {answer[i] ? answer[i] : `\xa0`}
        </span>
      );
    });
  };

  const displayLinesWithTagsForWords = (titleArray, correctAnswerIndex, updatedWordLength) => {
    return titleArray.map((preposition, index) => {
      let items = [];
      if ((correctAnswerIndex == 0 && index == correctAnswerIndex) || index == correctAnswerIndex) {
        items.push(tagPositioning(updatedWordLength));
      } else {
        items.push([
          <span
            className={`inline-block ${inputBoxConstants.prepositionClass(typingLength)}`}
            key={index}
            style={inputBoxConstants.prepositionStyle(index)}
          >
            {preposition}
          </span>
        ]);
      }
      return items;
    });
  };

  const tagPositioning = (updatedWordLength) => {
    if (tagIdentifier.tagType === 'start') {
      return (
        <Fragment>
          {displayTagValues()}
          {[...Array(updatedWordLength)].map((a, i) => {
            return (
              <span
                className={`inline-block ${inputBoxConstants.inputBoxClass(typingLength)}`}
                key={i}
                style={inputBoxConstants.inputBoxStyle(wordLength)}
              >
                {answer[i] ? answer[i] : `\xa0`}
              </span>
            );
          })}
        </Fragment>
      );
    }
    if (tagIdentifier.tagType === 'end') {
      return (
        <Fragment>
          {[...Array(updatedWordLength)].map((a, i) => {
            return (
              <span
                className={`inline-block ${inputBoxConstants.inputBoxClass(typingLength)}`}
                key={i}
                style={inputBoxConstants.inputBoxStyle(wordLength)}
              >
                {answer[i] ? answer[i] : `\xa0`}
              </span>
            );
          })}
          {displayTagValues()}
        </Fragment>
      );
    }
    if (tagIdentifier.tagType === 'middle') {
      return (
        <Fragment>
          {displayPrevTagLines()}
          {displayTagValues()}
          {displayAfterTagLines()}
        </Fragment>
      );
    }
  };

  const displayTagValues = () => {
    return tagValues.map((t, i) => {
      return (
        <span
          className={`inline-block ${inputBoxConstants.inputBoxClass(typingLength)}`}
          style={inputBoxConstants.tagStyle(wordLength)}
          key={i}
        >
          {t.value}
        </span>
      );
    });
  };

  const displayPrevTagLines = () => {
    let lines = [];
    for (let index = 0; index < typingArray.length; index++) {
      if (index > tagIdentifier.prevIndex) break;

      lines.push(
        <span
          className={`inline-block ${inputBoxConstants.inputBoxClass(typingLength)}`}
          key={index}
          style={inputBoxConstants.inputBoxStyle(wordLength)}
        >
          {answer[index] ? answer[index] : `\xa0`}
        </span>
      );
    }
    return lines.map((line, i) => {
      return line;
    });
  };

  const displayAfterTagLines = () => {
    let lines = [];
    let ansCounter = tagIdentifier.prevIndex + 1;
    let wordsCount = typing.split(' ').length;
    let wordArray = (wordsCount == 1) ? typing.split('') : correctAnswer.split('');
    for (let index = 0; index < wordArray.length; index++) {
      if (index > tagIdentifier.prevIndex + tagIdentifier.tagTotal) {
        lines.push(
          <span
            className={`inline-block ${inputBoxConstants.inputBoxClass(typingLength)}`}
            key={index}
            style={inputBoxConstants.inputBoxStyle(wordLength)}
          >
            {answer[ansCounter] ? answer[ansCounter] : `\xa0`}
          </span>
        );
        ansCounter++;
      }
    }
    return lines.map((line, i) => {
      return line;
    });
  };

  const displayItems = () => {
    const titleArray = typing.toLowerCase().split(' ');
    const correctAnswerIndex = titleArray.indexOf(correctAnswer.toLowerCase());

    // For phrases and words that has no tags and has no prepositions
    if ((typing.toLowerCase() === correctAnswer.toLowerCase() && learningType == 'word') && !tagIdentifier) {
      return displayLines();
    }

    const updatedWordLength = (tagIdentifier) ? wordLength - tagIdentifier.tagTotal : 0;
    // For phrases that have tags
    if (tagIdentifier && learningType == 'phrase') {
      return tagPositioning(updatedWordLength);
    }
    // For words that have tags && prepositions OR tags only (1 word or more)
    if (tagIdentifier && learningType == 'word') {
      return displayLinesWithTagsForWords(titleArray, correctAnswerIndex, updatedWordLength);
    }

    // For words that has no tags but has prepositions
    return titleArray.map((preposition, index) => {
      let items = [];
      let order =
        index === 0 ? 'first' : index === titleArray.length - 1 ? 'last' : 'between';

      if ((correctAnswerIndex == 0 && index == correctAnswerIndex) || index == correctAnswerIndex) {
        items.push(displayLines(index, correctAnswer.length));
      } else {
        items.push([
          <span
            className={`inline-block ${inputBoxConstants.prepositionClass(typingLength)}`}
            key={index}
            style={inputBoxConstants.prepositionStyle(index, order)}
          >
            {preposition}
          </span>
        ]);
      }
      return items;
    });
  };

  return (
    <div className='align-center'>
      {displayItems()}
    </div>
  );
};

export default InputField;
