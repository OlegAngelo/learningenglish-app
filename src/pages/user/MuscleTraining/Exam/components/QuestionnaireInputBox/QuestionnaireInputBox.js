import React, { useState, useEffect, Fragment } from 'react';

import InputField from './InputField';
import CorrectAnswer from './CorrectAnswer';
import IncorrectAnswer from './IncorrectAnswer';
import inputBoxConstants from './inputBoxConstants';
import QuestionnaireFeedback from '../QuestionnaireFeedback';
import questionHelper from '../../../../../../utils/questionHelper';
import SoftKeyboard from '../../../../../../shared/SoftKeyboard';

import styles from './QuestionnaireInputBox.module.css';

const QuestionnaireInputBox = ({
  className = 'mt-px-30',
  color,
  textColor,
  wordLength,
  correctAnswer,
  setResponse,
  setSelected,
  correctTranslation,
  correctTranslationClass,
  withFeedback,
  questionId,
  setAnswers,
  resultMessage,
  checkAnswer,
  timerProps,
  setIsShowfooter,
  categories,
  response,
  hint,
  children,
  typing,
  learningType,
  tagIdentifier,
  tagValues,
  handleAnswer = () => {}
}) => {
  const [displayType, setDisplayType] = useState(inputBoxConstants.input.type);
  const [answer, setAnswer] = useState('');
  const [answerWithNoTag, setAnswerWithNoTag] = useState('');
  const correctAnswerArray = correctAnswer.split('');
  const [incorrectIndices, setIncorrectIndices] = useState([]);
  const [responseMsg, setResponseMsg] = useState('');

  const inputBoxComponents = {
    input: InputField,
    correct: CorrectAnswer,
    incorrect: IncorrectAnswer,
  };

  useEffect(() => {
    if(!response) {
      setDisplayType(inputBoxConstants.input.type);
      setAnswer('');
      setAnswerWithNoTag('');
      setIncorrectIndices([]);
      setResponseMsg('');
    }
  }, [response]);

  const DisplayField = inputBoxComponents[displayType];

  const getCompleteAnswer = (answer) => {
    let finalAnswer = '';
    let answerArray = answer.split('');
    answerArray = [...Array(wordLength - tagIdentifier.tagTotal)].map((e, i) => {
      let order = i;
      if (tagIdentifier.tagType === 'start') {
        order = tagIdentifier.tagTotal + i;
      }
      if (tagIdentifier.tagType === 'middle') {
        if (i > tagIdentifier.prevIndex) {
          order = tagIdentifier.tagTotal + i;
        }
      }
      return {
        order: order,
        value: answer[i] ? answer[i] : '',
      };
    });

    for (let i = 0; i < wordLength; i++) {
      let tagValue = tagValues.filter((tag) => tag.order == i);
      finalAnswer += (tagValue.length === 0)
        ? answerArray.filter((ans) => ans.order == i)[0].value
        : tagValue[0].value;
    }
    return finalAnswer;
  };

  const checkDifference = (answer) => {
    let finalAnswer = answer;
    const incorrects = [];
    
    if (tagIdentifier) {
      finalAnswer = getCompleteAnswer(answer);
      setAnswer(finalAnswer);
    }

    // This is to handle special space character `\xa0`
    const answerArray = finalAnswer.replace(/(?=\s)[^\r\n\t]/g, ' ').split('');

    correctAnswerArray.forEach((char, index) => {
      if (answerArray[index] == null) return;

      if (char.toLowerCase() !== answerArray[index].toLowerCase()) {
        incorrects.push(index);
      }
    });

    const isCorrect = answerArray.length === correctAnswerArray.length && incorrects.length === 0;
    const responseType = isCorrect ? 'correct' : 'incorrect';
    const response = inputBoxConstants[responseType].response;

    setIncorrectIndices(incorrects);
    setResponse(response);
    setDisplayType(responseType);
    setIsShowfooter(true);
    handleAnswer(finalAnswer);
    setSelected(finalAnswer);
    setAnswers(answers => [
      ...answers,
      questionHelper.generateQuestionResult(
        categories,
        questionId,
        finalAnswer,
        undefined,
        hint,
      ),
    ]);
  };

  const displayFieldProps = {
    wordLength,
    checkDifference,
    answer,
    incorrectIndices,
    setIsShowfooter,
    hint,
    typing,
    correctAnswer,
    learningType,
    tagIdentifier,
    tagValues,
    answerWithNoTag,
  };

  useEffect(() => {
    if (timerProps.seconds == 0) {
      setAnswer(correctAnswer);
      setIncorrectIndices([]);
      setResponse('Failed...');
      setDisplayType('incorrect');
      setResponseMsg('Failed...');
    }
  }, [timerProps.seconds]);

  useEffect(() => {
    if (displayType != 'input') {
      setResponseMsg(
        checkAnswer(displayType === 'correct', timerProps.seconds)
        ? resultMessage
        : 'Failed...'
      );
    }
  }, [displayType]);

  const getFeedbackFragment = (className = '') => {
    return (
      <Fragment>
        {!withFeedback || displayType === inputBoxConstants.input.type ? null : (
          <QuestionnaireFeedback
            className={`font-sf-pro-text ${className}`}
            correctAnswer={correctAnswer}
            bottomText={correctTranslation}
            answer={answer}
            bottomTextClass={correctTranslationClass}
            response={responseMsg}
            topText={typing}
          />
        )}
      </Fragment>
    );
  };

  const getInputBoxFragment = (params) => {
    return (
      <div
        style={{ backgroundColor: color, color: textColor }}
        className={`bg-basic-400 rounded text-14 h-px-88 -mb-2 mx-2 flex flex-col justify-center text-left focus:outline-none ${
          styles[`${displayType}Outline`]
        } ${styles.inputBox}`}
      >
        <div className="flex justify-center">
          <DisplayField {...displayFieldProps} />
        </div>
      </div>
    );
  };

  const getResponseMessageFragment = (params) => {
    return (
      <div className={categories != 'audio-typing' ? className : ''}>
        <div
          className={`${!response ? `hidden` : `visible`} text-24 text-center font-black ${styles.response}`}
          style={{ color: questionHelper.getResponseColor(responseMsg) }}
        >
          &nbsp;{responseMsg}
        </div>
      </div>
    );
  };

  return (
    <div className={className}>
      {children ? (
        <Fragment>
          {getFeedbackFragment()}
          {children}
          {getResponseMessageFragment()}
          {getInputBoxFragment()}
          {responseMsg === 'Try again!' ? getFeedbackFragment('mt-px-24') : ''}
        </Fragment>
      ) : (
        <Fragment>
          {getFeedbackFragment()}
          {getResponseMessageFragment()}
          {getInputBoxFragment()}
        </Fragment>
      )}

      {!responseMsg && (
        <div className={`${styles.softKeyboardWrapper} w-full`}>
          <SoftKeyboard
            answer={answer}
            setAnswer={setAnswer}
            wordLength={!tagIdentifier ? wordLength : wordLength - tagIdentifier.tagTotal}
            submitHandler={checkDifference}
            setAnswerWithNoTag={setAnswerWithNoTag}
            hasTagIdentifier={tagIdentifier ? true : false}
          />
        </div>
      )}
    </div>
  );
};

export default QuestionnaireInputBox;
