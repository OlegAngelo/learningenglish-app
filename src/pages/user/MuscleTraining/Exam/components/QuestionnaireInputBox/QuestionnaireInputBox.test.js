import 'jsdom-global/register';
import React, { useState as useStateMock } from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import QuestionnaireInputBox from './QuestionnaireInputBox';

configure({ adapter: new Adapter() });

describe('QuestionnaireInputBox Component', () => {
  let props = {
      color: 'bg-basic-400',
      textColor: '#141414',
      setResponse: jest.fn(),
      setAnswers: jest.fn(),
      setIsShowfooter: true,
      setSelected: jest.fn(),
      correctAnswer: '',
      timerProps: {
        colorName: jest.fn(),
        isDisappearedTimer: false,
        percentage: jest.fn(),
        seconds: 20,
        isShowCommentary: false,
      },
  };

  describe('render', () => {
    it('should render exactly from snapshot', () => {
      const wrapper = mount(  <QuestionnaireInputBox  {...props}/>)
      expect(wrapper).toMatchSnapshot()
    });
  });

  describe('props', () => {
    it('should pass down to component', () => {
      const wrapper = mount(<QuestionnaireInputBox {...props} />);
      const wrapperProps = wrapper.props();

      expect(wrapperProps.color).toEqual(props.color);
      expect(wrapperProps.textColor).toEqual(props.textColor);
      expect(wrapperProps.setResponse).toEqual(props.setResponse);
      expect(wrapperProps.setAnswers).toEqual(props.setAnswers);
      expect(wrapperProps.setIsShowfooter).toEqual(props.setIsShowfooter);
      expect(wrapperProps.setSelected).toEqual(props.setSelected);
    });
  });
  
});
