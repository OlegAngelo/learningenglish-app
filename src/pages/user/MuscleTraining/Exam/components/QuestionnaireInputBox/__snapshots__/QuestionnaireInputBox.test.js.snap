// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionnaireInputBox Component render should render exactly from snapshot 1`] = `ReactWrapper {}`;
